### PR TITLE
Webpack for electron build

### PR DESCRIPTION
**Pull request by chrisnojima**
Ok so this bundles up our js using webpack so we have all our our requires and dependencies figured out up front. This'll result in a small app and a much quicker runtime. I changed the npm run commands

- npm run hot-server: Starts a hot module loading server. This lets you change files and refresh the electron app and see changes. We can make this work automagically but i have that turned off for now
- npm run start-hot: Start the gui, hitting the hot server. The normal flow will be to run hot-server and start-hot while devving
- npm run build-dev: Makes a dev build of the bundles under /dist (hot-server does this build but keeps it in memory and serves from a local webserver
- npm run build-prod: Makes a prod build (NOT TESTED YET)
- npm run start-cold: Run the gui off the of files in dist and not the local webserver

Things that are different now:
- [x] Avoid __dirname. Paths should be relative to desktop's root
- [x] We pass query params to the html files we launch (hot=true, dev=false, component=tracker)

We could theoretically start doing some cool webpack specific things if we wanted (with custom loaders, etc). Left for a future discussion

### CODE REVIEW

File: desktop/app/main.js [hunk 1]
```diff
@@ -1,32 +1,29 @@
...
-const Window = require('./window')
-const splash = require('./splash')
-const installer = require('./installer')
-const app = require('electron').app
-const path = require('path')
-const showDevTools = require('../../react-native/react/local-debug.desktop').showDevTools
-const isDev = require('../../react-native/react/local-debug.desktop').isDev
-const shell = require('electron').shell
-const helpURL = require('../../react-native/react/constants/urls').helpURL
+import menubar from 'menubar'
```
Comment (chrisnojima) [inline desktop/app/main.js]: Yay we can be ES2015 in here now

File: desktop/app/main.js [hunk 2]
```diff
@@ -1,32 +1,29 @@
...
+import {helpURL} from '../../react-native/react/constants/urls'
 
 const appPath = app.getAppPath()
-const menubarIconPath = path.resolve(appPath, 'Icon.png')
+const menubarIconPath = path.resolve('./Icon.png')
+const hot = process.env.HOT === 'true'
 
 const mb = menubar({
-  index: `file://${__dirname}/../renderer/launcher.html#debug=${isDev}`,
+  index: `file://${path.resolve('./renderer/launcher.html')}?debug=${!!isDev}&hot=${!!hot}`,
```
Comment (chrisnojima) [inline desktop/app/main.js]: All our html knows about being hot or not

File: desktop/client.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+const electron = require('electron-prebuilt')
+
+try{
+  fs.mkdirSync('dist')
+} catch(i) {
+}
+
+const name = 'dist/main.hot.bundle.js'
+
+const file = fs.createWriteStream(name)
```
Comment (chrisnojima) [inline desktop/client.js]: This donwloads the bootstrap file from the hot server and launches

File: desktop/package.json [hunk 4]
```diff
@@ -23,7 +25,7 @@
     "immutable": "^3.7.5",
     "lodash": "^3.10.1",
     "material-ui": "^0.13.3",
-    "menubar": "^2.3.0",
+    "menubar": "Zougi/menubar#6c4908f1994483604f5197c9a48b019ccb393781",
```
Comment (chrisnojima) [inline desktop/package.json]: menubar needs to support newer electron

File: desktop/package.json [hunk 5]
```diff
@@ -49,19 +50,25 @@
     "browser-sync": "^2.10.0",
     "electron-compile": "^1.0.1",
     "electron-packager": "^5.1.0",
-    "electron-prebuilt": "^0.35.4",
-    "electron-rebuild": "^1.0.0",
+    "electron-prebuilt": "mafintosh/electron-prebuilt#d76ae84fcd974114aa4da2809b8442ad8b9d43c2",
```
Comment (chrisnojima) [inline desktop/package.json]: electron prebuilt has an issue with electron rebuild, this is a snapshot of master which fixes that issue

File: desktop/renderer/index.html [hunk 6]
```diff
@@ -1,28 +1,29 @@
...
-      })
-      require("./index")
-
-      var path = require('path')
-      var injector = require('../../react-native/react/native/inject-styles')
-      injector(document, path.resolve(__dirname, './style.css'))
-      injector(document, path.resolve(__dirname, '../node_modules/font-awesome/css/font-awesome.min.css'))
+      var script = document.createElement('script')
+      script.type = 'text/javascript'
+      script.src = getQueryVariable('hot') === 'true' ? 'http://localhost:4000/dist/index.bundle.js' : '../dist/index.bundle.js'
```
Comment (chrisnojima) [inline desktop/renderer/index.html]: inject the bundle or hot url

**Comment by chrisnojima** [comment]
@keybase/react-hackers take a peek

**Comment by chrisnojima** [comment]
replaced by https://github.com/keybase/client/pull/1507

File: desktop/app/main.js (hunk 2 above)
Comment (MarcoPolo) [inline desktop/app/main.js]: why does it matter?

File: desktop/renderer/index.html (hunk 6 above)
Comment (MarcoPolo) [inline desktop/renderer/index.html]: nvm this answers the question above

**Pull request closed without merging**